grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcHealthCheckService.java

```diff
@@ -0,0 +1,287 @@
+            return SERVICE_UNKNOWN_RESPONSE;
+        } else {
+            throw new IllegalArgumentException("Invalid status:" + status);
+        }
+    }
+
+    private final SettableHealthChecker serverHealth;
+    private final Set<ListenableHealthChecker> healthCheckers;
+    private final Map<String, ListenableHealthChecker> grpcServiceHealthCheckers;
+    private final NonBlockingIdentityHashMap<StreamObserver<HealthCheckResponse>, Entry<String, ServingStatus>>
+            watchers = new NonBlockingIdentityHashMap<>();
+    @Nullable
+    private Server server;
+    private boolean updating;
+
+    GrpcHealthCheckService(
+            Set<ListenableHealthChecker> healthCheckers,
+            Map<String, ListenableHealthChecker> grpcServiceHealthCheckers,
+            List<HealthCheckUpdateListener> updateListeners
+    ) {
+        serverHealth = new SettableHealthChecker(false);
+        if (!updateListeners.isEmpty()) {
+            addServerHealthUpdateListener(ImmutableList.copyOf(updateListeners));
+        }
+        this.healthCheckers = ImmutableSet.<ListenableHealthChecker>builder()
+                                          .add(serverHealth)
+                                          .addAll(healthCheckers)
+                                          .build();
+        this.grpcServiceHealthCheckers = grpcServiceHealthCheckers;
+        setInternalHealthUpdateListener(this.healthCheckers);
+        setInternalHealthUpdateListener(this.grpcServiceHealthCheckers.values());
+    }
+
+    @Override
+    public void check(HealthCheckRequest request, StreamObserver<HealthCheckResponse> responseObserver) {
+        final String service = request.getService();
+        final ServingStatus status = checkServingStatus(service);
+        if (status == ServingStatus.SERVICE_UNKNOWN) {
+            responseObserver.onError(Status.NOT_FOUND
+                                             .withDescription(String.format(
+                                                     "The service name(=%s) is not registered in this service",
+                                                     service))
+                                             .asRuntimeException());
+            return;
+        }
+        responseObserver.onNext(getHealthCheckResponse(status));
+        responseObserver.onCompleted();
+    }
+
+    @Override
+    public void watch(HealthCheckRequest request, StreamObserver<HealthCheckResponse> responseObserver) {
+        final String service = request.getService();
+        final ServingStatus status = checkServingStatus(service);
+        final HealthCheckResponse response = getHealthCheckResponse(status);
+        responseObserver.onNext(response);
+        watchers.put(responseObserver, Maps.immutableEntry(service, response.getStatus()));
+        ((ServerCallStreamObserver<HealthCheckResponse>) responseObserver).setOnCancelHandler(() -> {
+            watchers.remove(responseObserver);
```

Comment:
Shouldn't we call `responseObserver.onCompleted();` when the server shutdown?

Reply:
We need to send `ServingStatus.NOT_SERVING` and close the active streams in `serverStopping(Server server)` listener.

Reply:
We should do it.